core/coreapi/pin.go

```diff
@@ -194,123 +214,106 @@ func (p *pinInfo) Type() string {
 	return p.pinType
 }
 
-func (api *PinAPI) pinLsAll(typeStr string, ctx context.Context) ([]coreiface.Pin, error) {
-	pinCh, errCh := PinLsAll(ctx, typeStr, api.pinning, api.dag)
-
-	var pins []coreiface.Pin
-loop:
-	for {
-		select {
-		case p, ok := <-pinCh:
-			if !ok {
-				break loop
-			}
-			pins = append(pins, p)
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		}
-	}
-	err := <-errCh
-	if err != nil {
-		return nil, err
-	}
-
-	return pins, nil
+func (p *pinInfo) Err() error {
+	return p.err
 }
 
-// PinLsAll is an internal function for returning a list of pins
-func PinLsAll(ctx context.Context, typeStr string, pin pin.Pinner, dag ipld.DAGService) (chan coreiface.Pin, chan error) {
-	ch := make(chan coreiface.Pin, 32)
-	errCh := make(chan error, 1)
+func (api *PinAPI) pinLsAll(typeStr string, ctx context.Context) <-chan coreiface.Pin {
+	out := make(chan coreiface.Pin)
 
 	keys := cid.NewSet()
-	AddToResultKeys := func(keyList []cid.Cid, typeStr string) error {
+
+	AddToResultKeys := func(keyList []cid.Cid, typeStr string) {
 		for _, c := range keyList {
 			if keys.Visit(c) {
-				select {
-				case ch <- &pinInfo{
+				out <- &pinInfo{
```

Comment:
Unless I'm missing something, we still need to select on the context, right?

Reply:
Looks like it. Also, `ctx` should be the first parameter of `pinLsAll`.

Reply:
Should be good now.